Feature/crkrenn/improved makefile

crkrenn opened the pull request:
This makefile has 4 new targets. It also defines GIT_HASH automatically. 

```
stop                 Stop all Docker containers
clean                Remove all Docker images, containers, and volumes
hash                 Show current short hash
restart-full-rebuild Restart all Docker containers, rebuilding everything
```

`make restart-full-rebuild` restarts the polis containers twice. For some reason restarting once results in an error connecting to postgres. Restarting a second time solves this problem.

crkrenn committed:
updated makefile

crkrenn committed:
remove docker volumes too with make clean

crkrenn committed:
improved makefile

crkrenn committed:
improved restart-full-rebuild

metasoarous commented:
Thanks for the submission @crkrenn.

This is super helpful, but I do worry slightly about the potential for nuking the database accidentally with `clean`. For development, I'm not sure it's a huge deal, but for production it obviously would be. So it might be worth thinking a bit about how to safeguard this, and how the flow might look different in a production setting.

For now, having separate `clean` and `clean-ALL` commands (with `ALL` in caps to catch the eye and make it harder to accidentally nuke data) might be a reasonable solution.

Alternatively (or additionally), we could start focusing on how to differentiate between the dev/prod ops, and place the safeguards there (there's basically NEVER a time where we'd want to delete the db volume in prod, and admins having to sort out the docker commands to do this is probably the right amount of friction). This cold either be an env variable, or phony `prod` target which sets some env variable to fork behavior so you can do `make prod ...`, and then check in the `clean` definition to see what mode you're in before carrying out the action.

Happy to discuss this further.

metasoarous commented:
Oh... another realization here. Am I right that these commands clean/rebuild _all_ docker containers/images/votumes, including those from other projects? If so, we should definitely scope so that this isn't the case.

A little outside of scope here, but tangentially related: May be nice to think about how to tag container/image names with tags for features in flight so that multiple instances of the system could be run on a dev machine. This may be mostly a "me" problem, but I frequently have multiple clones of the codebase checked out at any given time, so that if I'm working on things in one checkout, I don't have to mess around with stashing or anything in order to review and test out PRs. Would be nice to additionally not have to stop one set of docker processes to fire up and test out another.

crkrenn commented:
Thanks! I definitely need to work on my notification processing.

crkrenn committed:
responding to metasaurus's feedback

crkrenn committed:
new draft using tags and labels

crkrenn committed:
removed container names

crkrenn committed:
Merge branch 'edge' into feature/crkrenn/improved_makefile

crkrenn commented:
@metasoarous, I have addressed some of your concerns. This version uses docker names, tags, and labels to allow different branches to be associated with different docker images, volumes and containers. And, the rm commands should only nuke the appropriate resources. 

And, you can quickly switch back and forth between two deployed branches:
```
make stop
# edit tag in .env (TAG=yale)
make start
make stop
# edit tag in .env (TAG=edge)
make start
```
(base) crkrenn@admins-MacBook-Pro-2 crkrenn % docker images
REPOSITORY                  TAG       IMAGE ID       CREATED          SIZE
compdem/polis-file-server   yale      f5776ea2a306   10 seconds ago   116MB
compdem/polis-server        yale      49ae4d49193a   30 seconds ago   1.04GB
compdem/polis-postgres      yale      243c4dbd9fc4   44 seconds ago   184MB
compdem/polis-file-server   edge      5eed47794e9c   18 minutes ago   116MB
compdem/polis-server        edge      3b0489b5247b   19 minutes ago   1.04GB
compdem/polis-postgres      edge      6329a593a59d   19 minutes ago   184MB
<none>                      <none>    7f0fb5a4b3a8   23 minutes ago   116MB
<none>                      <none>    5d2a579feb23   27 minutes ago   1.04GB
compdem/polis-math          yale      d0c44a169885   27 minutes ago   718MB
compdem/polis-math          edge      bde31a3d7868   27 minutes ago   718MB
compdem/polis-nginx-proxy   yale      44b669f53380   28 minutes ago   22.1MB
<none>                      <none>    93957919ffa9   28 minutes ago   184MB
compdem/polis-nginx-proxy   edge      842034bf34eb   28 minutes ago   22.1MB
(base) crkrenn@admins-MacBook-Pro-2 crkrenn % docker volumes
docker: 'volumes' is not a docker command.
See 'docker --help'
(base) crkrenn@admins-MacBook-Pro-2 crkrenn % docker volume ls
DRIVER    VOLUME NAME
local     edge_backups
local     edge_postgres
local     yale_backups
local     yale_postgres
```

crkrenn committed:
added polis prefix to compose_project_name

crkrenn committed:
Merge branch 'edge' into feature/crkrenn/improved_makefile

crkrenn committed:
Update Makefile

crkrenn committed:
added prod targets

crkrenn commented:
I will try labeling containers with git branch name to avoid conflicts with tag.

crkrenn committed:
moved prod.env

crkrenn committed:
removed detach

crkrenn committed:
trying PROD option

crkrenn committed:
added 'PROD' option

crkrenn committed:
fixed spaces/tab bug

metasoarous committed:
Correct `env.prod` -> `env.prod`

crkrenn committed:
set makefile shell to bash

Commit:
fixed(?) PROD target

Commit:
cleaned up TAG processing

Commit:
strip whitespace from TAG definition

metasoarous committed:
Add compose yaml file config differentiation to makefile

metasoarous committed:
Clarify docstrings around PROD make target

metasoarous committed:
rename restart-FULL-REBUILD to start-FULL-REBUILD

For consistency with start-rebuild
